pkg/vali/validation.go

```diff
@@ -0,0 +1,99 @@
+package vali
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strings"
+)
+
+var mailRe = regexp.MustCompile(`\A[\w+\-.]+@[a-z\d\-]+(\.[a-z]+)*\.[a-z]+\z`)
+
+const tagName = "validate"
+
+type Validator interface {
+	Validate(interface{}) (bool, error)
+}
+type DefaultValidator struct {
+}
+
+func (v DefaultValidator) Validate(val interface{}) (bool, error) {
+	return true, nil
+}
+
+type StringValidator struct {
+	Min int
+	Max int
+}
+
+func (v StringValidator) Validate(val interface{}) (bool, error) {
+	l := len(val.(string))
+	if l == 0 {
+		return false, fmt.Errorf("cannot be blank")
+	}
+	if l < v.Min {
+		return false, fmt.Errorf("should be at least %v chars long", v.Min)
+	}
+	if v.Max >= v.Min && l > v.Max {
+		return false, fmt.Errorf("should be less than %v chars long", v.Max)
+	}
+	return true, nil
+}
+
+type NumberValidator struct {
+	Min int
+	Max int
+}
+
+func (v NumberValidator) Validate(val interface{}) (bool, error) {
+	num := val.(int)
+	if num < v.Min {
+		return false, fmt.Errorf("should be greater than %v", v.Min)
+	}
+	if v.Max >= v.Min && num > v.Max {
+		return false, fmt.Errorf("should be less than %v", v.Max)
+	}
+	return true, nil
+}
+
+type EmailValidator struct {
+}
+
+func (v EmailValidator) Validate(val interface{}) (bool, error) {
+	if !mailRe.MatchString(val.(string)) {
+		return false, fmt.Errorf("is not a valid email address")
+	}
+	return true, nil
+}
+func getValidatorFromTag(tag string) Validator {
+	args := strings.Split(tag, ",")
+	switch args[0] {
+	case "number":
+		validator := NumberValidator{}
+		fmt.Sscanf(strings.Join(args[1:], ","), "min=%d,max=%d", &validator.Min, &validator.Max)
+		return validator
+	case "string":
+		validator := StringValidator{}
+		fmt.Sscanf(strings.Join(args[1:], ","), "min=%d,max=%d", &validator.Min, &validator.Max)
+		return validator
+	case "email":
+		return EmailValidator{}
+	}
+	return DefaultValidator{}
+}
+func ValidateStruct(s interface{}) []error {
+	errs := []error{}
+	v := reflect.ValueOf(s)
```

Comment:
Why don't models define behaviour if they are valid or not? Using struct tags will require from you to use reflection, which will slow down the program just to check definitions. Wouldn't it make more sense to just call for example `account.Valid()` instead of passing it to ValidStruct function?
For generic validation package, this might be a good usecase, but I just can't see value in having it like this. We are trading the speed of execution for convenience of saving few more lines of code. Does not seem like a good tradeoff to me :relaxed: